merlin/schema/schema.py

```diff
@@ -82,37 +82,47 @@ def __post_init__(self):
         object.__setattr__(self, "tags", TagSet(self.tags))
         object.__setattr__(self, "dtype", md.dtype(self.dtype or md.unknown))
 
+        existing_shape = self.shape if self.shape.dims is not None else None
+        dims_shape = Shape(dims) if dims else None
+
+        value_counts = self.properties.get("value_count", {})
+        value_count = Domain(**value_counts)
+
         # Validate the allowed range of value count
-        value_count = Domain(**self.properties.get("value_count", {}))
         if value_count.min == 0 or value_count.max == 0:
             raise ValueError(
                 "`value_count` min and max must be greater than zero. "
                 f"Provided min: {value_count.min} max: {value_count.max}"
             )
 
-        if self.is_list is None:
-            object.__setattr__(self, "is_list", bool(value_count.max and value_count.max > 0))
+        counts_shape = (
+            Shape(((0, None), (value_count.min or 0, value_count.max))) if value_counts else None
+        )
 
-        if self.is_ragged is None:
-            if value_count.is_bounded and value_count.max > value_count.min:
-                object.__setattr__(self, "is_ragged", True)
-            elif value_count.is_bounded and value_count.max == value_count.min:
-                object.__setattr__(self, "is_ragged", False)
-            else:
-                object.__setattr__(self, "is_ragged", bool(self.is_list))
+        is_list = self.properties.pop("is_list", None)
+        is_ragged = self.properties.pop("is_ragged", None)
 
-        if self.is_ragged and not self.is_list:
-            raise ValueError(
-                "`is_ragged` is set to `True` but `is_list` is not. "
-                "Only list columns can set the `is_ragged` flag."
-            )
+        list_shape = Shape(((0, None), (0, None))) if is_list else None
 
-        if self.is_ragged and value_count.is_bounded and value_count.min == value_count.max:
-            raise ValueError(
-                "`is_ragged` is set to `True` but `value_count.min` == `value_count.max`. "
-                "If value_count min/max are equal. "
-                "This is a fixed size list and `is_ragged` should be set to False. "
-            )
+        new_shape = dims_shape or counts_shape or existing_shape or list_shape
+
+        self._validate_counts(new_shape, value_counts, is_ragged)
+        self._validate_flags(new_shape, is_list, is_ragged)
+
+        if new_shape:
+            object.__setattr__(self, "dtype", self.dtype.with_shape(new_shape))
+
+    @property
+    def shape(self):
+        return self.dtype.shape
+
+    @property
+    def is_list(self):
+        return self.dtype.shape.is_list
+
+    @property
+    def is_ragged(self):
+        return self.dtype.shape.is_ragged
```

Comment:
We might want to add `is_fixed`/`is_variable` here too, so that you don't have to reach through the dtype and the shape to access them (`col_schema.dtype.shape.is_variable`)